sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/Schema.java

```diff
@@ -0,0 +1,378 @@
+    /**
+     * Returns optional extra metadata.
+     */
+    @Nullable public abstract byte[] getMetadata();
+    abstract FieldTypeDescriptor.Builder toBuilder();
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setType(TypeName typeName);
+      abstract Builder setComponentType(@Nullable FieldTypeDescriptor componentType);
+      abstract Builder setRowSchema(@Nullable Schema rowSchema);
+      abstract Builder setMetadata(@Nullable byte[] metadata);
+      abstract FieldTypeDescriptor build();
+    }
+
+    /**
+     * Create a {@link FieldTypeDescriptor} for the given type.
+     */
+    public static FieldTypeDescriptor of(TypeName typeName) {
+      return new AutoValue_Schema_FieldTypeDescriptor.Builder().setType(typeName).build();
+    }
+
+    /**
+     * For container types, adds the type of the component element.
+     */
+    public FieldTypeDescriptor withComponentType(@Nullable FieldTypeDescriptor componentType) {
+      if (componentType != null) {
+        checkArgument(getType().isContainerType());
+      }
+      return toBuilder().setComponentType(componentType).build();
+    }
+
+    /**
+     * For ROW types, sets the schema of the row.
+     */
+    public FieldTypeDescriptor withRowSchema(@Nullable Schema rowSchema) {
+      if (rowSchema != null) {
+        checkArgument(getType().isCompositeType());
+      }
+      return toBuilder().setRowSchema(rowSchema).build();
+    }
+
+    /**
+     * Returns a copy of the descriptor with metadata sert set.
+     */
+    public FieldTypeDescriptor withMetadata(@Nullable byte[] metadata) {
+      return toBuilder().setMetadata(metadata).build();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (!(o instanceof FieldTypeDescriptor)) {
+        return false;
+      }
+      FieldTypeDescriptor other = (FieldTypeDescriptor) o;
+      return Objects.equals(getType(), other.getType())
+          && Objects.equals(getComponentType(), other.getComponentType())
+          && Objects.equals(getRowSchema(), other.getRowSchema())
+          && Arrays.equals(getMetadata(), other.getMetadata());
+
+    }
+
+    @Override
+    public int hashCode() {
+      return Arrays.deepHashCode(
+          new Object[] {getType(), getComponentType(), getRowSchema(), getMetadata()});
+    }
+  }
+
+
+  /**
+   * Field of a row. Contains the {@link FieldTypeDescriptor} along with associated metadata.
+   *
+   */
+  @AutoValue
+  public abstract static class Field implements Serializable {
+    /**
+     * Returns the field name.
+     */
+    public abstract String getName();
+
+    /**
+     * Returns the field's description.
+     */
+    public abstract String getDescription();
+
+    /**
+     * Returns the fields {@link FieldTypeDescriptor}.
+     */
+    public abstract FieldTypeDescriptor getTypeDescriptor();
+
+    /**
+     * Returns whether the field supports null values.
+     */
+    public abstract Boolean getNullable();
```

Comment:
`boolean isNullable()` should work as well.
But do we need it at the moment? In my opinion we should remove it until we support it to avoid confusion

Reply:
Yeah, because I plan on adding BigQuery bindings right after this, and that requires Nullable. I tried isNullable initially BTW, and for some reason could not get AutoValue to recognize it.